### ISSUE TITLE
Style/StringConcatenation changes escape sequences unexpectedly

### ISSUE THREAD

**Issue by adfoster-r7** [rubocop/rubocop#9543]
## Input file

I expected that the `Style/StringConcatenation` auto-correct wouldn't change any characters with the following input:

```ruby
$ cat example.rb 

query << "\x0a" + 'local-ddns'
query << "\x0b" + 'hmac-sha256' 
domain = "\x06" + Rex::Text.rand_text_alphanumeric(6)
org = "\x03" + Rex::Text.rand_text_alphanumeric(3)
```

We intentionally want these bytes left as-is for readability purposes when dealing with binary strings. Unfortunately it seems that the current Rubocop implementation changes these characters

### Current behavior

The strings are updated to use string interpolation / concatenation, but it's also changes the escape sequences are changed:

```ruby
query << "\nlocal-ddns"
query << "\vhmac-sha256"
domain = "\u0006#{Rex::Text.rand_text_alphanumeric(6)}"
org = "\u0003#{Rex::Text.rand_text_alphanumeric(3)}"
```
## Expected behavior

String interpolation is now used, but the escape sequences are unchanged:

```ruby
query << "\x0alocal-ddns"
query << "\x0bhmac-sha256"
domain = "\x06#{Rex::Text.rand_text_alphanumeric(6)}"
org = "\x03#{Rex::Text.rand_text_alphanumeric(3)}"
```

If the functionality of `Style/StringConcatenation` was restricted to purely string concatenation that would be great, and if there was a separate rule for replacing `\0x0a` to `\n` etc that would also be great as it could be disabled individually

## Extra example

When creating this issue I noticed that comments that exist at the end of strings are also changed:

```ruby
payload = "\x12" + # size
          "\x34" + # eip
          "\x56"   # etc
```

Becomes:

```ruby
payload = "\u00124V"    # etc
```

## Steps to reproduce the problem

```
$ cat example
query << "\nlocal-ddns"
query << "\vhmac-sha256"
domain = "\u0006#{Rex::Text.rand_text_alphanumeric(6)}"
org = "\u0003#{Rex::Text.rand_text_alphanumeric(3)}"

$ rubocop -A example.rb
```

Additional example:
https://github.com/rapid7/metasploit-framework/blob/master/modules/auxiliary/admin/scada/multi_cip_command.rb#L118-L122

## RuboCop version

```
1.10.0 (using Parser 3.0.0.0, rubocop-ast 1.4.1, running on ruby 2.7.2 x86_64-darwin19)
```

**Comment by dvandersluis** [comment]
Looks like this is an issue (not sure if it's a bug or intended) with the [https://github.com/whitequark/parser](parser gem):

```
$ ruby-parse -e '"\x0a" + "test"'
(send
  (str "\n") :+
  (str "test"))
```

When the node is parsed, `\x0a` is already converted to `\n`, so rubocop cannot change it back because we don't have the original value.

**Comment by adfoster-r7** [comment]
Thanks for taking a look!

Knowing that it's within the parser itself coincidentally helped point out why it was swapping over to unicode escape sequences

Most of our codebase sets the encoding:

```
$ cat foo.rb
# -*- coding: binary -*-

"\x0a"
"\x06"
$ ruby-parse foo.rb
(begin
  (str "\n")
  (str "\x06"))
```

But some of the files were missing it: 

```
$ cat foo.rb 
"\x0a"
"\x06"
$ ruby-parse foo.rb
(begin
  (str "\n")
  (str "\u0006"))
```

That doesn't solve the problem of being converted to `"\n"`, but it was still interesting to see

**Comment by adfoster-r7** [comment]
I'm really not sure at what point things would need to be handled to have this working for my particular use case:

```
$ irb
2.7.2 :001 > "\x0a"
 => "\n" 
2.7.2 :002 > "\x0b"
 => "\v"
2.7.2 :003 > require 'ripper'; require 'pp'; pp Ripper.sexp("'\x0a'")
[:program,
 [[:string_literal, [:string_content, [:@tstring_content, "\n", [1, 1]]]]]]
 => [:program, [[:string_literal, [:string_content, [:@tstring_content, "\n", [1, 1]]]]]] 
```

**Comment by cjlarose** [comment]
I don't thing this is necessarily a bug in the parser gem. Ruby Strings are just a byte array together with an encoding. When the `parser` gem sees "\x0a", it correctly parses the byte and sets the encoding. Later, if you `inspect` the String object (which doesn't have knowledge of the source code of the literal that was used to create it), it picks a representation, in this case `\n`.

The `parser` gem correctly preserves the source code literal, even if the `String` data structure doesn't.

```ruby
[15] pry(main)> node = Parser::CurrentRuby.parse('"\x0a" + "test"').loc
{
           :dot => nil,
      :selector => #<Parser::Source::Range (string) 7...8>,
           :end => nil,
         :begin => nil,
    :expression => #<Parser::Source::Range (string) 0...15>
}
[16] pry(main)> node.expression.source
"\"\\x0a\" + \"test\""
```

So it seems possible, in principle, for rubocop to use this information when correcting violations.
